### ISSUE TITLE
[WP][UX] Use admin-theme for Log In form

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#4410]
**Description of the need**

There should be a setting to use the admin theme for the log-in page, similar to how we have a checkbox to use the admin theme for the content edit form.

This would prevent everyone from needing to theme the log in page specially, when the log in form is only used for administrators.

**Proposed solution**

* Add a checkbox
* Switch the log-in / password reset / registration pages to use the selected theme
* Make sure these pages look good in the Seven theme

**Related issues**

* https://github.com/backdrop/backdrop-issues/issues/1528 Allow a layout to be selected on multiple paths
* https://github.com/backdrop/backdrop-issues/issues/4476 Add a "login pages" visibility condition

**Blocking issue**
* https://github.com/backdrop/backdrop-issues/issues/4487 Remove the 'view administration theme' permission

**Advocate**

Jen Lampton

---

PR: https://github.com/backdrop/backdrop/pull/3147

**Comment by jenlampton** [comment]
This feature was requested by @quicksketch in the latest dev meeting, but I think it's a great idea and would like to advocate for this feature in core.

**Comment by jenlampton** [comment]
I've pushed a PR for this, but there's one thing that's a little weird. You need to grant the `Anonymous` role access to `View administration theme` in order for this to work. That's a strange thing, since the log-in form is only ever accessable by Anonymous users. 

I'm going to add a work-around so that permission isn't necessary here.

Also, the pages need a little css cleanup so they look pretty :)

**Comment by jenlampton** [comment]
Okay, I've pushed an updated PR that adds an exception for the login pages. 
Oddly, it still renders the front-end layout even though when I check it's correctly assigned `admin_default`.

**Comment by jenlampton** [comment]
Okay, I found the problem.

```php
  // Safety check that we don't show an admin layout to a user who cannot view
  // the admin theme. This may happen because the cached layout list does not
  // include a permissions check on the admin theme.
  $acess_admin_theme = user_access('view the administration theme');
  if ($selected_layout->name === 'admin_default' && !$acess_admin_theme && !user_login_admin_theme()) {
    $selected_layout = layout_load('default');
  }
```

There's a chunk of code here that needs to be removed. This ties the display of the admin LAYOUT to the permission to view the admin THEME. The admin layout should always be rendered based on whether the page is an admin page or not -- regardless of which theme is in use on admin pages, or whether someone has access to view an administrative theme.

The PR removes this code.

**Comment by jenlampton** [comment]
Log in page: 
<img width="575" alt="login" src="https://user-images.githubusercontent.com/397895/82626295-4ef36980-9b9c-11ea-93a1-c1f47a8b32c1.png">

PW reset page:
<img width="583" alt="password" src="https://user-images.githubusercontent.com/397895/82626304-53b81d80-9b9c-11ea-82b0-5ec36757c035.png">

New account page:
<img width="585" alt="create" src="https://user-images.githubusercontent.com/397895/82626316-59156800-9b9c-11ea-94e9-112a4a662dc0.png">

**Comment by klonos** [comment]
There's 1000s of test errors in the PR for both php5 and php7, and when trying to access the sandbox, you get a 404 with `The requested page "/" could not be found.`

**Comment by jenlampton** [comment]
The new sandbox is working (though, tests are still failing) you can at least test it now :)

http://3147.backdrop.backdrop.qa.backdropcms.org/user/login
http://3147.backdrop.backdrop.qa.backdropcms.org/user/password

**Comment by docwilmot** [comment]
Looks great but no reset pwd link on the login page?

**Comment by jenlampton** [comment]
Ah, I forgot to put that back!  Marking nw :) ty!

**Comment by klonos** [comment]
Quick feedback:

- It's "Username or e-mail" in the login page; but its "Username or e-mail address" in the password reset page. Can we please drop the "address" bit?
- I would like a link to be able to return to the home page. There is a breadcrumb in `/user/password`, but there is non in `/user/login`.
- It seems a bit odd that on wider screens the breadcrumb is not aligned close to the page title and form elements:
  ![Screen Shot 2020-05-23 at 5 07 15 am](https://user-images.githubusercontent.com/2423362/82719441-50fc1c00-9cb3-11ea-8938-f8e978b93140.png)

**Comment by olafgrabienski** [comment]
For reference, below is also a screenshot of `user/register` page. I guess usually that page won't be made accessible on sites with a log in form to be only used for administrators, but it should however be considered.

![backdrop-pr-3147-user-register](https://user-images.githubusercontent.com/14188307/82793487-45565400-9e71-11ea-83a3-58de7984f811.png)

**Comment by jenlampton** [comment]
Yes, I included this screenshot [with all the others](https://github.com/backdrop/backdrop-issues/issues/4410#issuecomment-632449036) because it made sense that if the only people logging into a site were administrators, then the only people registering would also be administrators, so the style should match.

**Comment by jenlampton** [comment]
> no reset pwd link on the login page?

Added.

> It's "Username or e-mail" in the login page; but its "Username or e-mail address" in the password reset page. Can we please drop the "address" bit?

I found more instances of "Email address" in the UI, so I went the other way instead and added "address" were it was missing. 

> I would like a link to be able to return to the home page. There is a breadcrumb in /user/password, but there is none in /user/login.

hm, that's weird...  I don't know why there's no crumb here...  digging deeper.

> It seems a bit odd that on wider screens the breadcrumb is not aligned close to the page title and form elements

The page is still limited to a maximum width by layouts, so it shouldn't ever be too far away from the form. (The breadcrumb is in the same place on all pages in seven)

**Comment by jenlampton** [comment]
> hm, that's weird... I don't know why there's no crumb here... digging deeper.

Okay. The `user/login` page is defined as a `MENU_DEFAULT_LOCAL_TASK` which means that the other pages, `user/password` and `user/register` will show up "below" it in the menu structure, adding the `Log in` link into the breadcrumb on those pages (which is, I think, what we want)

However, these default tasks also don't include their parents in the breadcrumb trail, so that we don't get duplicate crumbs. The login page is special because it's parent is not the same as itself...  so I added an exception into `menu_get_active_breadcrumb()`

**Comment by docwilmot** [comment]
Missing `user/register` now, no?

**Comment by olafgrabienski** [comment]
Missing `user/register` now, no?

Depends on the settings of the website, should be available if registration of anonymous users is allowed.

**Comment by docwilmot** [comment]
I meant on this PR sandbox. Its missing.

**Comment by olafgrabienski** [comment]
Hm, works for me, now. (The setting was okay. I've cleared all caches.)

**Comment by docwilmot** [comment]
Ahh, no, misunderstanding: I meant there is no link to that page from the login page.

**Comment by olafgrabienski** [comment]
I see, good point!

**Comment by klonos** [comment]
I like to see this in the next release 🙂

**Comment by jenlampton** [comment]
I've updated the PR to pull the tabs off the page and add them as links in a more generic way, so that any module that adds a tab to the user pages will not loose the link because of this new setting.

PR needs review and testing again.

**Comment by klonos** [comment]
I'm reviewing this now @jenlampton ...@herbdool had left a comment in the PR re typos in a variable + the word "specail" in comments.

**Comment by klonos** [comment]
@jenlampton it mostly works as expected, thanks for working on this 🙏 

The only thing that I found annoying is that after I've enabled the setting, I had to manually clear caches before see the change take effect. Even after clearing caches, jumping from one `/user/*` page to another via the links provided, I would still end up in the page being rendered in Basis. I would then need to refresh the page once (w/o clearing caches) to get the proper theme.

The same odd behavior happened when I disabled the checkbox: I would still get Seven, until manually clearing caches. Then while on the "Log in" tab (switched back to Basis), I'd clock on either the "Reset password" tab, or the "Create new account" tab, and still get Seven 🤔 ...if I'd refresh the page once (w/o cache clear), I'd get back to Basis as expected.

Not sure if there's some weird browser or Backdrop caching going on there 🤷‍♂️

**Comment by olafgrabienski** [comment]
I also had to clear caches and refresh the page, but only once. Switching to "Reset password" or "Create account" using the links didn't need another cache clear.

Apart of that, the PR works for me in the sandbox but I saw a notice in DBlog:

> Undefined index: #active in seven_get_tabs_as_links() (line 227 of /home/qa/github/backdrop/backdrop/pr/3147/core/themes/seven/template.php).

**Comment by jenlampton** [comment]
I think it's reasonable to clear the caches after making this change. I'll add that to the submit handler :)

**Comment by docwilmot** [comment]
Somehow I feel a bit uncomfortable about this PR. I feel we're bending the theme and layout system a bit much here. At the very least the title of this issue is not accurate: actually we're forcing the default admin layout on to the login pages, and modifying the expected output of that layout as well. Those paths are inheriting the admin theme because we're setting those paths as admin paths. 

Note that any blocks added to the admin default layout will also end up on the login page. Is that OK? Not sure.

It looks like we're trying to build a page that looks different from both the default and the admin default pages. But the ideal way to have a page that looks the way you want is to use a different layout. Or bypass the layout system and use custom theme templates to generate the page, like the maintenance theme system does.

I dont have a solution yet, just my late opinion.

**Comment by herbdool** [comment]
@docwilmot you make some good points. I think it should be a new layout at least so it makes it easier to understand. I doubt people putting blocks on the admin pages would consider the login pages as part of that.

**Comment by jenlampton** [comment]
> we're forcing the default admin layout on to the login pages

Yep, the admin layout is used everywhere the admin theme is used.

> Note that any blocks added to the admin default layout will also end up on the login page. Is that OK? Not sure.

Hm. This does make it slightly too easy for someone to build an insecure view (an all content view that has no filters, for example), or a block containing text for admin-eyes only, then place it on the admin layout thinking that's a safe place, and getting a nasty surprise...

> It looks like we're trying to build a page that looks different from both the default and the admin default pages.

No, not really. I was trying to build a page that looks the same as the admin pages, but with the forms minimized and cleaned up as login pages usually are. This cleanup design decision is a special case for login pages, and is similar to what I add to all my front-end themes that end up getting used on login pages.

> But the ideal way to have a page that looks the way you want is to use a different layout.

Wait, you'd choose to add a new layout, even when there is no architectural reason to have one, just to make a page *look different*? I see that as the job of the theme. To me, adding a layout would be unnecessary overhead, creating more burden on site architects. (The next time you need to make a change to your header block settings, for example, you now need to do it in two places instead of one.)

A one-column layout works great for these log-in pages, and we were already using the one-column layout for the admin pages. 

I do see what you are saying about adding blocks to the admin layout and being surprised that they show up here, but I have another solution for that (see below).

> I think it should be a new layout at least so it makes it easier to understand.

Does it? I would think that having 2 one-column admin layouts with identical block placements is going to be confusing, and makes things unnecessarily complicated. (adding needs feedback label to see what others think...)

I also don't think that having special blocks on the log-in pages (or on admin pages that are NOT the log in pages) is an 80% use-case we need to ship with out of the box? 

What if we added an additional visibility condition so that people who do want a special layout for log-in pages can have one. We could also use it to safely place blocks with a "Not on the login pages" rule to keep them hidden. We could even add this rule automatically when new blocks are placed into the default admin layout :)

**Comment by herbdool** [comment]
> What if we added an additional visibility condition so that people who do want a special layout for log-in pages can have one. We could also use it to safely place blocks with a "Not on the login pages" rule to keep them hidden. We could even add this rule automatically when new blocks are placed into the default admin layout :)

I'm not comfortable with this suggestion. It seems like a lot of "magic" to get around a problem, something that is not easily understandable to the user. Plus it means updating every site's layouts by adding a visibility condition to all existing blocks in the admin layout as well.

**Comment by klonos** [comment]
>Note that any blocks added to the admin default layout will also end up on the login page. Is that OK? Not sure.

No it's not OK. I've missed that when reviewing 😞 

>>But the ideal way to have a page that looks the way you want is to use a different layout.
> Wait, you'd choose to add a new layout, even when there is no architectural reason to have one, just to make a page look _different_? I see that as the job of the theme. To me, adding a layout would be unnecessary overhead, creating more burden on site architects.

I'm with @docwilmot on this one. Yes, the theme would dictate what that page looks like, but the layout should be what specifies which blocks are added to the page, and in which regions.

Having said that, taking a look at the default admin layout, I see that the only blocks added to it is the breadcrumb and the main page content. These are exactly what we need on the login page, so it does seem like unnecessary duplication. I see what @jenlampton is saying, but the risk of exposing admin-eyes-only blocks to non-admins overweighs that.

Moreover, having a separate layout for the login page allows for the breadcrumb block to be placed in the main content area (as opposed to the rest of the admin page, where it's placed in the header region). That effectively solves the issue with he breadcrumb being "disconnected" from the rest of the login form, which [I've reported in my original review](https://github.com/backdrop/backdrop-issues/issues/4410#issuecomment-632969727):

![Screen Shot 2020-07-13 at 5 47 48 pm](https://user-images.githubusercontent.com/2423362/87318638-47fc2e00-c531-11ea-89a3-75caffe718fe.png)

...and the best thing about it is that I was able to achieve that without touching any code, via the Layout UI.

> Those paths are inheriting the admin theme because we're setting those paths as admin paths.

Yeah, this sounds wrong from a security point.

> We could also use it to safely place blocks with a "Not on the login pages" rule to keep them hidden.

+1 to adding this visibility condition.

> We could even add this rule automatically when new blocks are placed into the default admin layout :)

This one though sounds like "magic"; so -1 from me.

> Or bypass the layout system and use custom theme templates to generate the page, like the maintenance theme system does.

That does sound like a safe and "easy" option. Although appealing to go this way, it would remove the ability for non-coder site builders to easily place custom blocks on the login page. I feel that we _should_ add a dedicated layout, which is not "magic".

**Comment by klonos** [comment]
>+1 to adding this visibility condition (could be a follow-up).

Ha! @jenlampton took care of that 😄 : #4476

**Comment by jenlampton** [comment]
In today's dev meeting we discussed some of the other issues that were made evident by this proposed change. I've linked them in the top post, but also wanted to put them here:

**Related issues**

* https://github.com/backdrop/backdrop-issues/issues/1528 Allow a layout to be selected on multiple paths
* https://github.com/backdrop/backdrop-issues/issues/4487 Remove the 'view administration theme' permission

**Comment by BWPanda** [comment]
I heard about this issue in today's dev meeting, so thought to help out with a review. I hadn't heard anything about this issue before coming here, so I first read everything in the issue itself top to bottom. Then I tested what my dev site looked like normally (to give myself some context), then applied Jen's patch and tested that. Here're my thoughts:

- I don't personally like the admin theme for these pages as it looks too empty (there's not even an indication of *what* website you're logging into (i.e. no site name, logo, etc.)), but I appreciate that this is an opt-in feature (it *is* opt-in, right?)

- I think the form with the setting for this feature would look better like this:
  ![image](https://user-images.githubusercontent.com/2385329/90215086-f0a5f200-de3d-11ea-97b2-ededbb94201b.png)
  (i.e. convert the individual 'checkbox' fields into a single 'checkboxes' field, and change the wording of the new setting)

- Even before testing the PR or looking at the code, just by reading through this issue, I started feeling like this was an awful lot of weird changes just to get some pages styled in a different theme. I was therefore pleased to hear that @docwilmot expressed a [similar opinion](https://github.com/backdrop/backdrop-issues/issues/4410#issuecomment-657289079), as did @herbdool. Things that I'm referring to include:
  - "add a work-around so that permission isn't necessary here"
  - "adds an exception for the login pages"
  - "a chunk of code here that needs to be removed"
  - "where the layout is switched based on permission to view a theme ... I have removed that check too"
  - "added an exception into `menu_get_active_breadcrumb()`"
  - "pull the tabs off the page and add them as links in a more generic way"
  
  In my opinion, all of these little changes add up to a situation that makes me feel overwhelmed in terms of reviewing, and trying to work out if all of these changes are: 1. applicable to this PR, and 2. not going to have negative effects elsewhere...
  I'd suggest, where possible, seperate issues should be opened where something else needs changing for this to go ahead. Then they can be discussed and analysed on their own merit, rather than being lumped togeher into one big PR.

- There are still some comments above and/or in the PR that need addressing (e.g. an 'Undefined index' notice), so marking this as NW.

**Comment by klonos** [comment]
> convert the individual 'checkbox' fields into a single 'checkboxes' field

Good catch @BWPanda 👍 

> I started feeling like this was an awful lot of weird changes just to get some pages styled in a different theme. ... Things that I'm referring to include:

Wondering f adding a new core layout for these pages would make all or some of these changes unnecessary 🤔

**Comment by jenlampton** [comment]
Since the blocking issue didn't get in to 1.17 and is now scheduled for 1.18, bumping this milestone to 1.19. Let's discuss in today's meeting if we can get them both into 1.18?

**Comment by stpaultim** [comment]
For the record, my understand from the meeting yesterday (@quicksketch ??) was the we COULD possibly get both into 1.18. It was suggested that this might be easier if we get the blocking issue committed early in the release cycle.

**Comment by jenlampton** [comment]
Yeah, there are a lot of separate issues than need to be addressed.

Getting the view adminstration theme permission out of core is the first
step, which is why that issue is a blo keep for this one.

After that we still have a few more issues to sort out.

Adding a new context for "login pages" is also an interesting idea, and
would make the suggestion @klonos made to have a custom layout for these
pages possible.

**Comment by klonos** [comment]
With [Claro](https://www.drupal.org/project/claro) and [Gin](https://www.drupal.org/project/gin) admin themes gaining traction in Drupal-land, I wanted to mention how the https://www.drupal.org/project/gin_login module also changes the login screen, based on any of those themes. Here's a couple of screenshots showing regular/dark mode:

![image](https://user-images.githubusercontent.com/2423362/99159574-1dd89980-26e6-11eb-84ff-9c574f9bd766.png)

![image](https://user-images.githubusercontent.com/2423362/99159577-24ffa780-26e6-11eb-9c97-7ae5409957e3.png)

**Comment by klonos** [comment]
...the reason why I mentioned the above is because I would like us to use the same style as install/update.php for our login pages (these are still part of the Seven admin theme):

![image](https://user-images.githubusercontent.com/2423362/99159957-af4a0a80-26ea-11eb-8fb4-4cdb84958fc8.png)

**Comment by BWPanda** [comment]
I still think we need some indication of what site you're logging into (e.g. the site logo or site name).

**Comment by klonos** [comment]
Yeah, we could change the page title to "Log into [site:name]"

**Comment by jenlampton** [comment]
Hm, I'm not sure I like including the design changes here. Can we tackle that in a follow-up issue?

**Comment by klonos** [comment]
Sure @jenlampton 👍🏼 ...can you please rebase the PR so that we get a fresh sandbox + have the GHA tests run on it?

I will then move the design-related comments in a separate follow-up issue.

**Comment by stpaultim** [comment]
> Since the blocking issue didn't get in to 1.17 and is now scheduled for 1.18, bumping this milestone to 1.19. Let's discuss in today's meeting if we can get them both into 1.18

I just re-read the entire conversation and it's not clear to me what the blocking issue is or whether it has been resolved yet (despite the fact that I myself referred to a blocking issue in one comment). 

Is this issue still blocked? And if so, by what?

**Comment by jenlampton** [comment]
The blocker was https://github.com/backdrop/backdrop-issues/issues/4487

**Comment by herbdool** [comment]
Yes. Closing this.
